File: LeetCode/Medium/1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.py
class Solution:
    def nearestExit(self, maze: List[List[str]], entrance: List[int]) -> int:
        m, n = len(maze), len(maze[0])
        start_row, start_col = entrance
        queue: deque[tuple[int, int, int]] = deque([(start_row, start_col, 0)])
        directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        maze[start_row][start_col] = "+"

        while queue:
            row, col, steps = queue.popleft()

            if row in (0, m - 1) or col in (0, n - 1):
                if row != start_row or col != start_col:
                    return steps

                if maze[row][col] != "+":
                    return steps

            for dx, dy in directions:
                new_row, new_col = row + dx, col + dy

                if (
                    0 <= new_row < m
                    and 0 <= new_col < n
                    and maze[new_row][new_col] == "."
                ):
                    maze[new_row][new_col] = "+"
                    queue.append((new_row, new_col, steps + 1))

        return -1

File: LeetCode/Medium/1926-nearest-exit-from-entrance-in-maze/1926-nearest-exit-from-entrance-in-maze.rs
impl Solution {
    pub fn nearest_exit(maze: Vec<Vec<char>>, entrance: Vec<i32>) -> i32 {
        let (m, n) = (maze.len() as i32, maze[0].len() as i32);
        let (start_row, start_col) = (entrance[0], entrance[1]);
        let mut maze = maze;
        maze[start_row as usize][start_col as usize] = '+';
        let mut queue = std::collections::VecDeque::from(vec![(start_row, start_col, 0)]);

        while let Some((row, col, steps)) = queue.pop_front() {
            if row == 0 || row == m - 1 || col == 0 || col == n - 1 {
                if row != start_row || col != start_col {
                    return steps;
                }

                if maze[row as usize][col as usize] != '+' {
                    return steps;
                }
            }

            for (dx, dy) in [(1, 0), (0, 1), (-1, 0), (0, -1)] {
                let (new_row, new_col) = (row + dx, col + dy);

                if (0 <= new_row && new_row < m)
                    && (0 <= new_col && new_col < n)
                    && maze[new_row as usize][new_col as usize] == '.'
                {
                    maze[new_row as usize][new_col as usize] = '+';
                    queue.push_back((new_row, new_col, steps + 1));
                }
            }
        }

        return -1;
    }
}
